File: utils/src/main/java/jetbrains/exodus/core/dataStructures/persistent/PersistentLong23TreeSet.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.core.dataStructures.persistent;

import jetbrains.exodus.core.dataStructures.hash.LongIterator;
import org.jetbrains.annotations.Nullable;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class PersistentLong23TreeSet implements PersistentLongSet {

    private final Persistent23Tree<PersistentLongMap.Entry<Boolean>> set;

    public PersistentLong23TreeSet() {
        this(null);
    }

    private PersistentLong23TreeSet(@Nullable final AbstractPersistent23Tree.RootNode<PersistentLongMap.Entry<Boolean>> root) {
        set = new Persistent23Tree<>(root);
    }

    @Override
    public PersistentLongSet.ImmutableSet beginRead() {
        return new ImmutableSet(set.getRoot());
    }

    @Override
    public PersistentLong23TreeSet getClone() {
        return new PersistentLong23TreeSet(set.getRoot());
    }

    @Override
    public PersistentLongSet.MutableSet beginWrite() {
        return new MutableSet(set);
    }

    protected static class ImmutableSet extends PersistentLong23TreeMap.ImmutableMap<Boolean> implements PersistentLongSet.ImmutableSet {

        ImmutableSet(RootNode<PersistentLongMap.Entry<Boolean>> root) {
            super(root);
        }

        @Override
        public boolean contains(long key) {
            return containsKey(key);
        }

        @Override
        public LongIterator longIterator() {
            return new IteratorImpl(iterator());
        }

        @Override
        public LongIterator reverseLongIterator() {
            return new IteratorImpl(reverseIterator());
        }

        @Override
        public LongIterator tailLongIterator(long key) {
            return new IteratorImpl(tailEntryIterator(key));
        }

        @Override
        public LongIterator tailReverseLongIterator(long key) {
            return new IteratorImpl(tailReverseEntryIterator(key));
        }
    }

    protected static class MutableSet implements PersistentLongSet.MutableSet {

        private final PersistentLong23TreeMap.MutableMap<Boolean> map;

        MutableSet(Persistent23Tree<PersistentLongMap.Entry<Boolean>> set) {
            map = new PersistentLong23TreeMap.MutableMap<>(set);
        }

        @Override
        public LongIterator longIterator() {
            return new IteratorImpl(map.iterator());
        }

        @Override
        public LongIterator reverseLongIterator() {
            return new IteratorImpl(map.reverseIterator());
        }

        @Override
        public LongIterator tailLongIterator(long key) {
            return new IteratorImpl(map.tailEntryIterator(key));
        }

        @Override
        public LongIterator tailReverseLongIterator(long key) {
            return new IteratorImpl(map.tailReverseEntryIterator(key));
        }

        @Override
        public boolean isEmpty() {
            return map.isEmpty();
        }

        @Override
        public int size() {
            return map.size();
        }

        @Override
        public boolean contains(long key) {
            return map.get(key) == Boolean.TRUE;
        }

        @Override
        public void add(long key) {
            map.put(key, Boolean.TRUE);
        }

        @Override
        public boolean remove(long key) {
            Boolean result = map.remove(key);
            if (result == null) {
                return false;
            }
            return result;
        }

        @Override
        public void clear() {
            map.setRoot(null);
        }

        @Override
        public boolean endWrite() {
            return map.endWrite();
        }
    }

    protected static class IteratorImpl implements LongIterator {
        private final Iterator<PersistentLongMap.Entry<Boolean>> it;

        IteratorImpl(Iterator<PersistentLongMap.Entry<Boolean>> it) {
            this.it = it;
        }

        @Override
        public long nextLong() {
            PersistentLongMap.Entry<Boolean> entry = it.next();
            if (entry == null) {
                throw new NoSuchElementException();
            }
            return entry.getKey();
        }

        @Override
        public boolean hasNext() {
            return it.hasNext();
        }

        @Override
        public Long next() {
            PersistentLongMap.Entry<Boolean> entry = it.next();
            if (entry == null) {
                throw new NoSuchElementException();
            }
            return entry.getKey();
        }

        @Override
        public void remove() {
            throw new UnsupportedOperationException();
        }
    }
}

File: utils/src/main/java/jetbrains/exodus/core/dataStructures/persistent/PersistentLong23TreeSet.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.core.dataStructures.persistent

import jetbrains.exodus.core.dataStructures.hash.LongIterator

class PersistentLong23TreeSet private constructor(root: AbstractPersistent23Tree.RootNode<PersistentLongMap.Entry<Boolean?>?>?) :
    PersistentLongSet {
    private val set: Persistent23Tree<PersistentLongMap.Entry<Boolean?>?>

    constructor() : this(null)

    init {
        set = Persistent23Tree(root)
    }

    override fun beginRead(): PersistentLongSet.ImmutableSet {
        return ImmutableSet(set.root)
    }

    override val clone: PersistentLongSet
        get() = PersistentLong23TreeSet(set.root)

    override fun beginWrite(): PersistentLongSet.MutableSet {
        return MutableSet(set)
    }

    protected class ImmutableSet internal constructor(root: RootNode<PersistentLongMap.Entry<Boolean?>?>?) :
        PersistentLong23TreeMap.ImmutableMap<Boolean?>(root), PersistentLongSet.ImmutableSet {
        override fun contains(key: Long): Boolean {
            return containsKey(key)
        }

        override fun longIterator(): LongIterator {
            return IteratorImpl(iterator())
        }

        override fun reverseLongIterator(): LongIterator {
            return IteratorImpl(reverseIterator())
        }

        override fun tailLongIterator(key: Long): LongIterator {
            return IteratorImpl(tailEntryIterator(key))
        }

        override fun tailReverseLongIterator(key: Long): LongIterator {
            return IteratorImpl(tailReverseEntryIterator(key))
        }
    }

    protected class MutableSet internal constructor(set: Persistent23Tree<PersistentLongMap.Entry<Boolean?>?>) :
        PersistentLongSet.MutableSet {
        private val map: PersistentLong23TreeMap.MutableMap<Boolean?>

        init {
            map = PersistentLong23TreeMap.MutableMap(set)
        }

        override fun longIterator(): LongIterator {
            return IteratorImpl(map.iterator())
        }

        override fun reverseLongIterator(): LongIterator {
            return IteratorImpl(map.reverseIterator())
        }

        override fun tailLongIterator(key: Long): LongIterator {
            return IteratorImpl(map.tailEntryIterator(key))
        }

        override fun tailReverseLongIterator(key: Long): LongIterator {
            return IteratorImpl(map.tailReverseEntryIterator(key))
        }

        override val isEmpty: Boolean
            get() = map.isEmpty

        override fun size(): Int {
            return map.size()
        }

        override fun contains(key: Long): Boolean {
            return map[key] === java.lang.Boolean.TRUE
        }

        override fun add(key: Long) {
            map.put(key, java.lang.Boolean.TRUE)
        }

        override fun remove(key: Long): Boolean {
            return map.remove(key) ?: return false
        }

        override fun clear() {
            map.root = null
        }

        override fun endWrite(): Boolean {
            return map.endWrite()
        }
    }

    protected class IteratorImpl internal constructor(private val it: Iterator<PersistentLongMap.Entry<Boolean?>?>?) :
        LongIterator {
        override fun nextLong(): Long {
            val entry = it!!.next() ?: throw NoSuchElementException()
            return entry.key
        }

        override fun hasNext(): Boolean {
            return it!!.hasNext()
        }

        override fun next(): Long {
            val entry = it!!.next() ?: throw NoSuchElementException()
            return entry.key
        }

        override fun remove() {
            throw UnsupportedOperationException()
        }
    }
}
